src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java

```diff
@@ -0,0 +1,262 @@
+
+            /* Produce is only given in the final stage */
+            if (bushComponent.currentStage == bushComponent.stages.length - 1) {
+                EntityRef produce;
+                /* Handle a dodgy produce */
+                try {
+                    produce = entityManager.create(bushComponent.produce);
+                    boolean giveSuccess = inventoryManager.giveItem(harvester, target, produce);
+                    if (!giveSuccess) {
+                        Vector3f position = event.getTargetLocation().add(0, 0.5f, 0);
+                        produce.send(new DropItemEvent(position));
+                        produce.send(new ImpulseEvent(random.nextVector3f(15.0f)));
+                    }
+                } catch (NullPointerException e) {
+                    logger.error("Unable to create produce: " + bushComponent.produce);
+                }
+                if (bushComponent.sustainable) {
+                    doBushGrowth(bushComponent, -1);
+                } else {
+                    entity.send(new DoDestroyPlant());
+                    worldProvider.setBlock(bushComponent.position, blockManager.getBlock(BlockManager.AIR_ID));
+                }
+                event.consume();
+            }
+        }
+    }
+
+    /**
+     * Handles the seed drop on plant destroyed event.
+     *
+     * @param event  The event corresponding to the plant destroy.
+     * @param entity Reference to the plant entity.
+     */
+    @ReceiveEvent
+    public void onBushDestroyed(CreateBlockDropsEvent event, EntityRef entity, BushDefinitionComponent bushComponent) {
+        entity.send(new DoDestroyPlant());
+        event.consume();
+    }
+
+    /**
+     * Grow the bush by one stage in a specified direction
+     *
+     * @param bushComponent The definition of the bush to grow
+     * @param direction     The direction to grow in. positive indicates forward & negative indicated backwards
+     */
+    private EntityRef doBushGrowth(BushDefinitionComponent bushComponent, int direction) {
+        bushComponent.currentStage += direction;
+        worldProvider.setBlock(bushComponent.position, bushComponent.stages[bushComponent.currentStage].block);
+        EntityRef newBush = blockEntityRegistry.getBlockEntityAt(bushComponent.position);
+        newBush.addOrSaveComponent(bushComponent);
+        return newBush;
+    }
+
+
+    /**
+     * Builds the list of growth stages from the prefab data
+     *
+     * @param growthStages The prefab GrowthStage data
+     * @return The array of growth stages
+     */
+    private GrowthStage[] buildGrowthStages(Map<String, GrowthStage> growthStages) {
+        Object[] values = growthStages.values().toArray();
+        Iterator<String> keys = growthStages.keySet().iterator();
+        GrowthStage[] stages = new GrowthStage[values.length];
+        for (int i = 0; i < values.length; i++) {
```

Comment:
have a look a `Map.entrySet() for this kind of iteration, you can access key and value from the same object if you iterate over entries`.

Also this looks like it depends on the order of the map which is (likely) not guaranteed for persisted components, better double check it here =)

Reply:
Order seems to be guaranteed for persisted components as all of my testing hasn't turned up an instance where it isn't.